clang/runtime/dpct-rt/include/sparse_utils.hpp

```diff
@@ -512,11 +513,33 @@ class sparse_matrix_desc {
   /// Get the uplo attribute
   /// \return uplo value
   std::optional<oneapi::mkl::uplo> get_uplo() const noexcept { return _uplo; }
+  /// Set the number of non-zero elements
+  /// \param nnz [in] The number of non-zero elements.
+  void set_nnz(std::int64_t nnz) { _nnz = nnz; }
+  /// Get the type of the value pointer.
+  /// \return The type of the value pointer.
+  library_data_t get_value_type() { return _value_type; }
 
 private:
   template <typename index_t, typename value_t> void set_data() {
+    void *row_ptr = nullptr;
+    if (!_row_ptr && !_temp_row_ptr) {
+      row_ptr = _temp_row_ptr = dpct::dpct_malloc(
+          sizeof(index_t) * (_row_num + 1), get_default_queue());
+    } else if (!_row_ptr && _temp_row_ptr) {
+      row_ptr = _temp_row_ptr;
+    } else if (_row_ptr && !_temp_row_ptr) {
+      row_ptr = _row_ptr;
+    } else if (_row_ptr && _temp_row_ptr) {
+      dpct::dpct_memcpy(
+          _row_ptr, _temp_row_ptr, sizeof(index_t) * (_row_num + 1),
+          memcpy_direction::device_to_device, get_default_queue());
+      row_ptr = _row_ptr;
+      dpct::dpct_free(_temp_row_ptr, get_default_queue());
+      _temp_row_ptr = nullptr;
+    }
```

Comment:
I see what you are trying to do here, but this is not allowed for oneMKL's sparse BLAS `matrix_handle_t`. You are not allowed to switch out arrays that you have already passed to `set_csr_data()` with new arrays. This is possibly why you had problems with `cusparseSpGEMM` to `matmat` portability support.
Although it _might_ work, it is not guaranteed to work the way you have set this up.

Reply:
See the reply in the 1st comment.